File: Algorithms/src/hard/Tree Inversions.py
import sys
import math
sys.setrecursionlimit(1 << 25)
data = sys.stdin.buffer.read().split()
it = iter(data)

def ni():
    return int(next(it))

def solve_case():
    n = ni()
    q = ni()

    # read node values (1-based)
    a_vals = [0] * (n + 1)
    for i in range(1, n + 1):
        a_vals[i] = ni()

    # graph
    g = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        x = ni(); y = ni()
        g[x].append(y)
        g[y].append(x)

    # prepare arrays for Euler tour and LCA
    # we'll determine LG from n
    LG = max(1, (n).bit_length() + 1)
    P = [[-1] * LG for _ in range(n + 1)]
    lvl = [0] * (n + 1)
    IN = [0] * (n + 1)
    OUT = [0] * (n + 1)
    ID = []  # 1-based indexing for positions; we'll add a dummy at front to align with C++ indices
    ID.append(0)  # dummy so ID[1] corresponds to first entry
    cur = 1

    def dfs(u, par):
        nonlocal cur
        # enter
        ID.append(u)
        IN[u] = cur
        cur += 1
        lvl[u] = lvl[par] + 1 if par != 0 else 1
        P[u][0] = par
        for i in range(1, LG):
            pp = P[u][i-1]
            P[u][i] = P[pp][i-1] if (pp != -1 and pp != 0) else (0 if pp != -1 else -1)
        for v in g[u]:
            if v == par: continue
            dfs(v, u)
        # exit
        ID.append(u)
        OUT[u] = cur
        cur += 1

    # run dfs from node 1, parent 0
    dfs(1, 0)
    # now cur is next unused index; valid positions are 1..cur-1
    total_positions = cur - 1

    # Fix P table: some entries may be 0 for parent 0; keep 0 as valid parent sentinel
    # (we used 0 for root parent, consistent with C++ using 0 in dfs)

    def lca(u, v):
        if lvl[u] < lvl[v]:
            u, v = v, u
        # lift u up to same level as v
        diff = lvl[u] - lvl[v]
        i = 0
        while diff:
            if diff & 1:
                u = P[u][i]
            diff >>= 1
            i += 1
        if u == v:
            return u
        # binary lift both
        for i in range(LG - 1, -1, -1):
            if P[u][i] != -1 and P[u][i] != P[v][i]:
                u = P[u][i]
                v = P[v][i]
        return P[u][0]

    def dis(u, v):
        w = lca(u, v)
        # number of nodes on path inclusive (same as C++: lvl[u] + lvl[v] - 2*lvl[w] + 1)
        return lvl[u] + lvl[v] - 2*lvl[w] + 1

    # BLOCK decomposition over Euler positions 1..total_positions
    sz = int(math.sqrt(total_positions)) or 1
    BL = [0] * (total_positions + 1)
    for i in range(1, total_positions + 1):
        BL[i] = (i - 1) // sz + 1

    # read queries and map to [l, r] on Euler tour like in C++ code
    class Qobj:
        __slots__ = ('id', 'l', 'r', 'lc')
        def __init__(self, _id, l, r, lc):
            self.id = _id; self.l = l; self.r = r; self.lc = lc

    queries = []
    for qi in range(q):
        x = ni(); y = ni()
        lc = lca(x, y)
        if IN[x] > IN[y]:
            x, y = y, x
        if lc == x:
            l = IN[x]; r = IN[y]
        else:
            l = OUT[x]; r = IN[y]
        queries.append(Qobj(qi, l, r, lc))

    # sort queries in Mo order (block of l, then r)
    queries.sort(key=lambda qq: (BL[qq.l], qq.r))

    # compress values of a to small range for freq array
    vals = sorted({a_vals[i] for i in range(1, n+1)})
    comp = {v:i+1 for i, v in enumerate(vals)}  # 1-based compressed values
    a = [0] * (n + 1)
    for i in range(1, n+1):
        a[i] = comp[a_vals[i]]
    max_val = len(vals) + 5

    # structures used by check()
    freq = [0] * (max_val)
    VIS = [0] * (n + 1)  # VIS[node] whether node currently included
    cur_sum = 0  # counts number of unordered equal-value pairs among included nodes (sum in C++)

    def check(node):
        # toggles node inclusion, updates cur_sum and freq
        nonlocal cur_sum
        val = a[node]
        # remove old contribution
        cur_sum -= (freq[val] * (freq[val] - 1)) // 2
        if VIS[node]:
            freq[val] -= 1
            VIS[node] = 0
        else:
            freq[val] += 1
            VIS[node] = 1
        # add new contribution
        cur_sum += (freq[val] * (freq[val] - 1)) // 2

    if q == 0:
        return []

    # initialize pointers like in C++: curL = Q[0].l, curR = Q[0].l - 1
    curL = queries[0].l
    curR = curL - 1
    answers = [0] * q

    for qq in queries:
        L = qq.l; R = qq.r
        while curL < L:
            check(ID[curL]); curL += 1
        while curL > L:
            curL -= 1; check(ID[curL])
        while curR < R:
            curR += 1; check(ID[curR])
        while curR > R:
            check(ID[curR]); curR -= 1

        xnode = ID[curL]; ynode = ID[curR]

        # if lca is not already inside the segment, include it temporarily
        if qq.lc != xnode and qq.lc != ynode:
            check(qq.lc)

        length = dis(xnode, ynode)
        # total unordered pairs among nodes on path = comb(len, 2) = len*(len-1)//2
        answers[qq.id] = (length * (length - 1)) // 2 - cur_sum

        if qq.lc != xnode and qq.lc != ynode:
            check(qq.lc)

    return answers

def main():
    T = ni()
    out_lines = []
    for _ in range(T):
        ans = solve_case()
        for v in ans:
            out_lines.append(str(v))
    sys.stdout.write("\n".join(out_lines))

if __name__ == "__main__":
    main()

File: Algorithms/src/hard/Tree Inversions.cpp
#include <bits/stdc++.h>
using namespace std;
#define int long long int
const int N=100005;
const int LG = 22;

int lvl[N], P[N][LG], a[N], in[N], out[N], cur = 1;
int ID[N << 1], BL[N << 1], freq[N], VIS[N], sum = 0;
vector<int>g[N];

void clear(int n) {
    cur = 1;
    sum = 0;
    for(int i = 0; i <= n; i++) {
        lvl[i] = 0, a[i] = 0, in[i] = 0, out[i] = 0, freq[i] = 0, VIS[i] = 0;
        g[i].clear();
        for(int j = 0; j < LG; j++) {
            P[i][j] = -1;
        }
    }
    for(int i = 0; i <= 2 * n; i++) {
        ID[i] = 0, BL[i] = 0;
    }
}

void dfs(int u, int par){
    ID[cur] = u;
    in[u] = cur++;
    lvl[u] = 1+lvl[par];
    P[u][0] = par;
    for(int i = 1; i < LG; i++) {
        if(P[u][i - 1] != -1) {
            P[u][i] = P[P[u][i - 1]][i - 1];
        }
    }
    for(int v:g[u]){
        if (v == par) continue;
        dfs(v, u);
    }
    ID[cur] = u;
    out[u] = cur++;
}
 
int lca(int u, int v){
    int i, lg;
    if (lvl[u] < lvl[v]) swap(u, v);
 
    for(lg = 0; (1<<lg) <= lvl[u]; lg++);
    lg--;
 
    for(i=lg; i>=0; i--){
        if (lvl[u] - (1<<i) >= lvl[v])
            u = P[u][i];
    }
 
    if (u == v) 
        return u;
 
    for(i=lg; i>=0; i--){
        if (P[u][i] != -1 and P[u][i] != P[v][i])
            u = P[u][i], v = P[v][i];
    }
 
    return P[u][0];
}

int dis(int u, int v){
    if (lvl[u] < lvl[v]) swap(u, v);
    int w = lca(u, v);
    return lvl[u] + lvl[v] - 2*lvl[w] + 1;
}

struct query{
    int id, l, r, lc;
    bool operator < (const query& rhs){
        return (BL[l] == BL[rhs.l]) ? (r < rhs.r) : (BL[l] < BL[rhs.l]);
    }
}Q[N];

void check(int x){
    sum -= (freq[a[x]] * (freq[a[x]] - 1)) / 2;
    if(VIS[x]) {
        freq[a[x]] -= 1;
    } else {
        freq[a[x]] += 1;
    }
    sum += (freq[a[x]] * (freq[a[x]] - 1)) / 2;
    VIS[x] ^= 1;
}


void solve() {

    int n, q;
    cin >> n >> q;

    clear(n);

    for(int i = 1; i <= n; i++) {
        cin >> a[i];
    }

    for(int i = 0; i < n - 1; i++) {
        int x, y;
        cin >> x >> y;
        g[x].push_back(y);
        g[y].push_back(x);
    }

    dfs(1, 0);

    int sz = sqrt(cur);
    for (int i = 1; i <= cur; i++) {
        BL[i] = (i - 1) / sz + 1;
    }

    for(int i = 0; i < q; i++) {
        int x, y;
        cin >> x >> y;
        Q[i].lc = lca(x, y);
        if (in[x] > in[y]) {
            swap(x, y);
        }
        if(Q[i].lc == x) Q[i].l = in[x], Q[i].r = in[y];
        else Q[i].l = out[x], Q[i].r = in[y];
        Q[i].id = i;
    }
    sort(Q, Q + q);

    vector<int> arr(q, 0);

    int curL = Q[0].l, curR = Q[0].l - 1;
    for (int i = 0; i < q; i++){
 
        while (curL < Q[i].l) check(ID[curL++]);
        while (curL > Q[i].l) check(ID[--curL]);
        while (curR < Q[i].r) check(ID[++curR]);
        while (curR > Q[i].r) check(ID[curR--]);
 
        int x = ID[curL], y = ID[curR];

        if (Q[i].lc != x and Q[i].lc != y) check(Q[i].lc);
        
        int len = dis(x, y);
        arr[Q[i].id] = (len * (len - 1)) / 2 - sum;
 
        if (Q[i].lc != x and Q[i].lc != y) check(Q[i].lc);
    }

    for(int i = 0; i < q; i++) {
        cout << arr[i] << endl;
    }

}


#undef int 
int main() {
#define int long long int
ios_base::sync_with_stdio(false); 
cin.tie(0); 
cout.tie(0);

    int T;
    cin >> T;

    for(int tc = 1; tc <= T; tc++){
        // cout << "Case #" << tc << ": ";
        solve();
    }

return 0;  

}
